src/rate/providers/currency-api.service.ts

```diff
@@ -0,0 +1,42 @@
+import { ICurrencyApiResponse } from '../interfaces/currency-api-response.interface';
+import { IHandler } from '../interfaces/handler.interface';
+import { IRate } from '../interfaces/rate.interface';
+import { Log } from '../logging/logger.decorator';
+import { CURRENCY_API_URL } from '../rate.constants';
+import { HttpService } from '@nestjs/axios';
+import { Injectable } from '@nestjs/common';
+import { firstValueFrom } from 'rxjs';
+
+@Injectable()
+export class CurrencyApiHandler implements IHandler {
+  private nextHandler: IHandler;
+
+  constructor(private readonly httpService: HttpService) {}
+
+  setNext(handler: IHandler): IHandler {
+    this.nextHandler = handler;
+    return handler;
+  }
+
+  @Log(CURRENCY_API_URL)
```

Comment:
Cool! I forgot about decorators in typescript/nest.js, and created just class for it )
I'll refactor my class to something like yours ))